### PR TITLE
Tabular: Added ImagePredictorModel

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

- Added ImagePredictorModel

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1041-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/3/index.html

File: core/src/autogluon/core/features/types.py [hunk 1]
```diff
@@ -31,6 +31,9 @@
 # feature is a generated feature based off of a text feature that is an ngram.
 S_TEXT_NGRAM = 'text_ngram'
 
+# feature is an object type that contains a string url to an image that can be utilized in computer vision
+S_IMAGE_URL = 'image_url'
```
Comment (jwmueller) [inline core/src/autogluon/core/features/types.py:35]: Might _path be more appropriate suffix than _url?

**Comment by jwmueller** [comment]
Can FeatureMetadata docs include somewhere: 

For descriptions of each special feature-type, see:  `autogluon.core.features.types`

File: core/src/autogluon/core/features/types.py (hunk 1 above)
Comment (jwmueller) [reply core/src/autogluon/core/features/types.py:35]: `image_url` is missing from list of possible special types in FeatureMetadata documentation

**Comment by jwmueller** [comment]
Unrelated to this PR but can FeatureMetadata docs include a basic usage example specifying metadata and then calling `fit()`?  Without this, the role of the class may be unclear to user.

File: features/src/autogluon/features/generators/auto_ml_pipeline.py [hunk 2]
```diff
@@ -42,6 +42,11 @@ class AutoMLPipelineFeatureGenerator(PipelineFeatureGenerator):
     enable_raw_text_features : bool, default False
         Whether to use the raw text features. The generated raw text features will end up with '_raw_text' suffix.
         For example, 'sentence' --> 'sentence_raw_text'
+    enable_vision_features : bool, default True
```
Comment (jwmueller) [inline features/src/autogluon/features/generators/auto_ml_pipeline.py:45]: Is this argument actually needed? why would a user specify `image_url` in metadata but not want to leverage the images?

File: core/src/autogluon/core/features/types.py (hunk 1 above)
Comment (jwmueller) [reply core/src/autogluon/core/features/types.py:35]: Niche use-case but may want to clarify:  how user can leverage both image and its file-path both as predictive features

File: tabular/src/autogluon/tabular/trainer/model_presets/presets.py [hunk 3]
```diff
@@ -81,6 +83,7 @@
     NNFastAiTabularModel: 'NeuralNetFastAI',
     TabTransformerModel: 'Transformer',
     TextPredictorModel: 'TextNeuralNetwork',
+    ImagePredictorModel: 'ImagePredictor',
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/trainer/model_presets/presets.py:86]: to standardize, should it be: `ImageNeuralNetwork` like for Text?

File: tabular/tests/unittests/models/test_image_predictor.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+
+def test_image_predictor(fit_helper):
+    train_data, _, test_data = ImageDataset.from_folders('https://autogluon.s3.amazonaws.com/datasets/shopee-iet.zip')
+    feature_metadata = FeatureMetadata.from_df(train_data).add_special_types({'image': ['image_url']})
+    predictor = TabularPredictor(label='label').fit(
+        train_data=train_data,
+        hyperparameters={'AG_IMAGE_NN': {'epochs': 2}},
+        feature_metadata=feature_metadata
+    )
```
Comment (jwmueller) [inline tabular/tests/unittests/models/test_image_predictor.py:14]: maybe also test setting `predictor=None` and reloading it before inference

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        idx_to_image_map = X[['image']]
+        idx_to_image_map = idx_to_image_map.reset_index(drop=True).reset_index(drop=False)
+        y_pred_proba = y_pred_proba.merge(idx_to_image_map, on='image')
+        y_pred_proba = y_pred_proba.drop(columns=['image'])
+        class_preds = []
+        for clss in self._classes:
+            cur_clss = y_pred_proba[y_pred_proba['id'] == clss]
+            cur_clss = cur_clss.set_index('index')['score']
+            class_preds.append(cur_clss)
+        y_pred_proba = pd.concat(class_preds, axis=1).to_numpy()
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: Does this post-processing really need to be so complex?

@zhreshold  Should the issue be fixed in autogluon.vision/gluonCV first before merging this PR?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        if X_val is not None:
+            X_val[self._label_column_name] = y_val
+
+        verbosity_image = max(0, verbosity - 1)
+        root_logger = logging.getLogger()
+        root_log_level = root_logger.level
+        # TODO: ImagePredictor doesn't use problem_type in any way at present.
+        #  It also doesn't error or warn if problem_type is not one it expects.
+        self.model = ImagePredictor(
+            problem_type=self.problem_type,
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:119]: should we raise clear error message if regression?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+# TODO: Add regression support
+# TODO: refit_full does not work as expected: It won't use all data, will just split train data internally.
+class ImagePredictorModel(AbstractModel):
+    """Wrapper of autogluon.vision.ImagePredictor"""
+    nn_model_name = 'image_nn'
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._label_column_name = 'label'
+        self._load_model = None  # Whether to load inner model when loading.
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:28]: why None instead of boolean?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
+
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:1]: what happens with HPO for this model?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+             y_val: Optional[pd.Series] = None,
+             time_limit: Optional[int] = None,
+             sample_weight=None,
+             verbosity=2,
+             **kwargs):
+        """The internal fit function
+
+        Parameters
+        ----------
+        X
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:78]: can doc be clearer about what X/X_val must look like?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+                       time_limit=time_limit,
+                       hyperparameters=params,
+                       random_state=0)
+        # self.model.set_verbosity(verbosity)  # TODO: How to set verbosity of fit predictor?
+        root_logger.setLevel(root_log_level)  # Reset log level
+
+    def _predict_proba(self, X, **kwargs):
+        X = self.preprocess(X, **kwargs)
+
+        if self.problem_type == REGRESSION:
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:143]: Should this error be moved earlier as well for `fit()`?

**Review by jwmueller — CHANGES_REQUESTED** [review]
Looks pretty good, mostly have minor suggestions. Major issue is the post-processing code needed at the end of `predict_proba()`

File: tabular/tests/unittests/models/test_image_predictor.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (jwmueller) [inline tabular/tests/unittests/models/test_image_predictor.py:1]: Can we have a similar script in `examples/`, so it can be shared with folks that want to test out this functionality?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 5 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: Ideally, gluoncv should just return the normal pred_proba format. I don't know why it is in the format it is, but it takes a **_lot_** of work to convert it to the standard format.

Comment (jwmueller) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: Right I guess @zhreshold can tell us what's best for now regarding this issue.

Comment (zhreshold) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: See #1044, if the post-processing isn't a blocker then you could remove these once #1044 get merged.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        root_log_level = root_logger.level
+        # TODO: ImagePredictor doesn't use problem_type in any way at present.
+        #  It also doesn't error or warn if problem_type is not one it expects.
+        self.model = ImagePredictor(
+            problem_type=self.problem_type,
+            path=self.path,
+            # eval_metric=self.eval_metric,  # TODO: Vision only works with accuracy
+            verbosity=verbosity_image
+        )
+        classes = list(y.unique())
```
Comment (sxjscience) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:124]: We may need to call the following to avoid the case that some rare classes only appear in the validation set:

```python
if X_val is None:
    classes = list(y.unique())
else:
    classes = list(pd.concat([y, y_val]).unique())
```

File: tabular/src/autogluon/tabular/trainer/model_presets/presets.py (hunk 3 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/trainer/model_presets/presets.py:86]: I think rather Text should be TextPredictor. This makes it clear that the model is coming from the other predictors within AutoGluon, rather than something custom and specific to Tabular. In future, I'll also add `TabularPredictorModel` as a model type.

File: features/src/autogluon/features/generators/auto_ml_pipeline.py (hunk 2 above)
Comment (Innixma) [reply features/src/autogluon/features/generators/auto_ml_pipeline.py:45]: possibly for debugging, I agree it isn't really necessary, but doesn't hurt.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 5 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: Will wait for #1044 to merge before addressing in this PR

File: tabular/tests/unittests/models/test_image_predictor.py (hunk 4 above)
Comment (Innixma) [reply tabular/tests/unittests/models/test_image_predictor.py:14]: That is tested elsewhere, since we know predictor doesn't persist models in memory be default.

File: core/src/autogluon/core/features/types.py (hunk 1 above)
Comment (Innixma) [reply core/src/autogluon/core/features/types.py:35]: renamed to `image_path`. Regarding using path as both text and image feature at the same time: Let's leave that for later, seems really niche.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 10 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:134]: added

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+
+
+# FIXME: Avoid hard-coding 'image' column name
+# TODO: Handle multiple image columns?
+# TODO: Handle multiple images in a single image column?
+# TODO: Add regression support
+# TODO: refit_full does not work as expected: It won't use all data, will just split train data internally.
+class ImagePredictorModel(AbstractModel):
+    """Wrapper of autogluon.vision.ImagePredictor"""
+    nn_model_name = 'image_nn'
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: renamed to `image_predictor`, this encompasses the entire predictor so it shouldn't be specific.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+        """The internal fit function
+
+        Parameters
+        ----------
+        X
+            Features of the training dataset
+        y
+            Labels of the training dataset
+        X_val
+            Features of the validation dataset
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: removed docs, this was copy/pasted from Tabular's text predictor model.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 12 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Good point, added

File: core/src/autogluon/core/features/types.py (hunk 1 above)
Comment (Innixma) [reply core/src/autogluon/core/features/types.py]: added `image_path` to FeatureMetadata docs

**Comment by Innixma** [comment]
> Can FeatureMetadata docs include somewhere:
> 
> For descriptions of each special feature-type, see: `autogluon.core.features.types`

Added

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 7 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Added extended description

**Comment by szha** [comment]
Job PR-1041-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/4/index.html

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 8 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:1]: Currently image model would fail unless a search space was provided by user (same as TextNN currently)

File: tabular/tests/unittests/models/test_image_predictor.py (hunk 11 above)
Comment (Innixma) [reply tabular/tests/unittests/models/test_image_predictor.py:1]: Would prefer to get a golden dataset example for multimodality, we currently don't have one besides petfinder and petfinder is somewhat large for an example.

**Comment by szha** [comment]
Job PR-1041-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/5/index.html

File: tabular/src/autogluon/tabular/trainer/model_presets/presets.py (hunk 3 above)
Comment (jwmueller) [reply tabular/src/autogluon/tabular/trainer/model_presets/presets.py:86]: @sxjscience Can we standardize this in the future?
I also think `TextPredictor` might be more clear name than `TextNeuralNetwork`.

Comment (sxjscience) [reply tabular/src/autogluon/tabular/trainer/model_presets/presets.py:86]: Sounds good to me.

**Comment by szha** [comment]
Job PR-1041-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/6/index.html

File: tabular/src/autogluon/tabular/trainer/model_presets/presets.py (hunk 3 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/trainer/model_presets/presets.py:84]: Updated to TextPredictor

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 5 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Code block is now removed, no longer necessary after #1044

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 8 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:1]: Added default HPO search space, but we probably want to have a better search space in future

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py (hunk 9 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Removed this, it was copy-paste from TextPredictorModel

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+    def get_memory_size(self) -> int:
+        """Return the memory size by calculating the total number of parameters.
+
+        Returns
+        -------
+        memory_size
+            The total memory size in bytes.
+        """
+        # FIXME: How to get correct size of model?
+        total_size = 0
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Thanks, added `len(pickle.dumps(self.model._classifier, pickle.HIGHEST_PROTOCOL))`

**Comment by szha** [comment]
Job PR-1041-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/9/index.html

**Comment by Innixma** [comment]
@zhreshold @sxjscience @jwmueller Addressed comments, ready for 2nd round of review.

File: features/src/autogluon/features/generators/auto_ml_pipeline.py [hunk 16]
```diff
@@ -42,6 +42,11 @@ class AutoMLPipelineFeatureGenerator(PipelineFeatureGenerator):
...
+        [Experimental]
+        Whether to keep 'object' features identified as 'image_path' special type. Features of this form should have a url path to an image file as their value.
```
Comment (jwmueller) [inline features/src/autogluon/features/generators/auto_ml_pipeline.py:47]: "url path" -> "string path"

File: core/src/autogluon/core/features/feature_metadata.py [hunk 17]
```diff
@@ -176,6 +177,17 @@ def keep_features(self, features: list, inplace=False):
         features_to_remove = [feature for feature in self.get_features() if feature not in features]
         return self.remove_features(features=features_to_remove, inplace=inplace)
 
+    def add_special_types(self, type_map_special: dict, inplace=False):
+        """Adds special types to features"""
```
Comment (jwmueller) [inline core/src/autogluon/core/features/feature_metadata.py:181]: might be helpful if docstring provides a basic example use-case.

**Review by jwmueller — APPROVED** [review]
looks good, just had minor remaining comments

**Review by sxjscience — APPROVED** [review]
I’ve tried and it works well.

File: core/src/autogluon/core/features/feature_metadata.py (hunk 17 above)
Comment (Innixma) [reply core/src/autogluon/core/features/feature_metadata.py]: added

File: features/src/autogluon/features/generators/auto_ml_pipeline.py (hunk 16 above)
Comment (Innixma) [reply features/src/autogluon/features/generators/auto_ml_pipeline.py]: updated

**Comment by szha** [comment]
Job PR-1041-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/10/index.html
